File: Creational Design Patterns/Factory Method/FactoryMethodExample.cpp
#include<bits/stdc++.h>
using namespace std;

// Document Interface
class Document{
    public:
        virtual void use()=0;
        virtual ~Document(){};
};

// Concrete Document Classes
class PDFDocument: public Document{
    public:
        void use() override{
            cout<<"You are using a PDF Document"<<endl;
        }
};
class WordDocument: public Document{
    public:
        void use() override{
            cout<<"You are using a Word Document"<<endl;
        }
};

// Creator Interface
class DocumentCreator{
    public:
        virtual Document* createDocument()=0;
        virtual ~DocumentCreator(){};
};

// Concrete Creator Classes
class PDFCreator: public DocumentCreator{
    public:
        Document* createDocument() override{
            return new PDFDocument();
        }
};
class WordCreator: public DocumentCreator{
    public:
        Document* createDocument() override{
            return new WordDocument();
        }
};



int main(){
    DocumentCreator* creator;
    creator=new PDFCreator();
    Document* document1=creator->createDocument();
    document1->use();

    creator=new WordCreator();
    Document*document2=creator->createDocument();
    document2->use();

    delete creator;
    delete document1;
    delete document2;

    return 0;
}

File: 1. Creational Design Patterns/Factory Method/Python/FactoryMethodExample.py
# Factory Method Pattern Example in Python
# Document processing with different document types

from abc import ABC, abstractmethod

class Document(ABC):
    @abstractmethod
    def create_content(self):
        pass

class PDFDocument(Document):
    def create_content(self):
        return "PDF document content created"

class WordDocument(Document):
    def create_content(self):
        return "Word document content created"

class DocumentCreator(ABC):
    @abstractmethod
    def create_document(self):
        pass
    
    def process_document(self):
        doc = self.create_document()
        return f"Processing: {doc.create_content()}"

class PDFCreator(DocumentCreator):
    def create_document(self):
        return PDFDocument()

class WordCreator(DocumentCreator):
    def create_document(self):
        return WordDocument()

def main():
    pdf_creator = PDFCreator()
    word_creator = WordCreator()
    
    print(pdf_creator.process_document())
    print(word_creator.process_document())

if __name__ == "__main__":
    main()
